persistent_volumes_test.go

```diff
@@ -0,0 +1,410 @@
+	}
+	pv.Spec.ClaimRef = claimRef
+
+	pv, err = testClient.PersistentVolumes().Create(pv)
+	if err != nil {
+		t.Fatalf("Unexpected error creating pv: %v", err)
+	}
+
+	waitForPersistentVolumePhase(testClient, pv.Name, w, api.VolumeBound)
+
+	pv, err = testClient.PersistentVolumes().Get(pv.Name)
+	if err != nil {
+		t.Fatalf("Unexpected error getting pv: %v", err)
+	}
+	if pv.Spec.ClaimRef == nil {
+		t.Fatalf("Unexpected nil claimRef")
+	}
+	if pv.Spec.ClaimRef.Namespace != claimRef.Namespace || pv.Spec.ClaimRef.Name != claimRef.Name {
+		t.Fatalf("Bind mismatch! Expected %s/%s but got %s/%s", claimRef.Namespace, claimRef.Name, pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)
+	}
+}
+
+// TestPersistentVolumeMultiPVs tests binding of one PVC to 100 PVs with
+// different size.
+func TestPersistentVolumeMultiPVs(t *testing.T) {
+	_, s := framework.RunAMaster(t)
+	defer s.Close()
+
+	deleteAllEtcdKeys()
+	testClient, _, binder := createClients(s)
+	binder.Run()
+	defer binder.Stop()
+
+	maxPVs := 100
+	pvs := make([]*api.PersistentVolume, maxPVs)
+	for i := 0; i < maxPVs; i++ {
+		// This PV will be claimed, released, and deleted
+		pvs[i] = createPV("pv-"+strconv.Itoa(i), "/tmp/foo"+strconv.Itoa(i), strconv.Itoa(i)+"G",
+			[]api.PersistentVolumeAccessMode{api.ReadWriteOnce}, api.PersistentVolumeReclaimDelete)
+	}
+
+	pvc := createPVC("pvc-2", strconv.Itoa(maxPVs/2)+"G", []api.PersistentVolumeAccessMode{api.ReadWriteOnce})
+
+	w, _ := testClient.PersistentVolumes().Watch(api.ListOptions{})
+	defer w.Stop()
+
+	for i := 0; i < maxPVs; i++ {
+		_, _ = testClient.PersistentVolumes().Create(pvs[i])
+	}
+
+	_, _ = testClient.PersistentVolumeClaims(api.NamespaceDefault).Create(pvc)
+
+	// wait until the binder pairs the claim with a volume
+	waitForAnyPersistentVolumePhase(w, api.VolumeBound)
+
+	// only one PV is bound
+	bound := 0
+	for i := 0; i < maxPVs; i++ {
+		pv, err := testClient.PersistentVolumes().Get(pvs[i].Name)
+		if err != nil {
+			t.Fatalf("Unexpected error getting pv: %v", err)
+		}
+		if pv.Spec.ClaimRef == nil {
+			continue
+		}
+		// found a bounded PV
+		p := pv.Spec.Capacity[api.ResourceStorage]
+		pvCap := p.Value()
+		p = resource.MustParse(strconv.Itoa(maxPVs/2) + "G")
+		pvcCap := p.Value()
+		if pv.Spec.ClaimRef.Name != pvc.Name || pvCap != pvcCap {
+			t.Fatalf("Bind mismatch! Expected %s capacity %v but got %s capacity %v", pvc.Name, resource.MustParse(strconv.Itoa(50)+"G"),
+				pv.Spec.ClaimRef.Name, pv.Spec.Capacity[api.ResourceStorage])
+		}
+		t.Logf("claim bounded to %s capacity %v", pv.Name, pv.Spec.Capacity[api.ResourceStorage])
+		bound += 1
+	}
+
+	if bound != 1 {
+		t.Fatalf("Only 1 PV should be bound but got %d", bound)
+	}
+
+	// deleting a claim releases the volume
+	if err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Delete(pvc.Name, nil); err != nil {
+		t.Errorf("error deleting claim %s", pvc.Name)
+	}
+
+	waitForAnyPersistentVolumePhase(w, api.VolumeReleased)
```

Comment:
Shouldn't this be waiting for the bound volume to be released, not just any volume?